pymc/distributions/distribution.py

```diff
@@ -83,6 +85,18 @@
 PLATFORM = sys.platform
 
 
+class MomentRewrite(GraphRewriter):
+    def add_requirements(self, fgraph):
+        fgraph.attach_feature(ReplaceValidate())
+
+    def apply(self, fgraph):
+        for node in fgraph.toposort():
+            if hasattr(node.op, "fgraph"):
+                self.rewrite(node.op.fgraph)
```

Comment:
We have to double check this isn't mutating the fgraph of the original Ops (I am not sure they are cloned by default in FunctionGraph).

Mote generally I am not sure this is the best approach. The fgraph is supposed to be just a property of the Op, the real underlying variables are op.outputs I think? So we could create a FunctionGraph from those, rewrite them and then create a new node with a new op that uses those new outputs and replace it in the outer graph...

However these Ops don't have a shared construction API IIRC, so we may need special logic for each one. They are Scan, OpFromGraph, Composite, and ScalarLoop. We can raise NotImplementedError for anything else to be future proof.

Reply:
@ricardoV94 thank you for review! I will try to replace `Scan` op in outer graph by new node with moment replaced inner graph.

Reply:
@ricardoV94 If I understand correctly, new `Scan` op in outer graph is already created in my implementation when I clone graph  from dist function. So we need only replace inner graph, or do I miss something?

Reply:
I mean something like
```python
if hasattr(node.op, "fgraph"):
    inner_graph = node.op.fgraph.clone()
    self.rewrite(inner_graph)
    node.op.fgraph = inner_graph
```
should be enough?

Reply:
Unfortunately we can't just modify the fgraph, as that's just a dynamic property of the Op. In some places it may be used directly, but that's not guaranteed. The Op inner inputs / outputs are the real properties of the Op, which means we need to create a new Scan op that uses the new outputs. Recreating a Scan is pretty annoying, you can check pytensor scan rewrites to see how it's approached there.